libs/api_common/device_helper.cpp

```diff
@@ -59,3 +68,108 @@ get_device_handle()
+            SetThreadpoolWait(async_ioctl_completion->wait, nullptr, nullptr);
+
+            // Close the wait.
+            CloseThreadpoolWait(async_ioctl_completion->wait);
+        }
+
+        if (async_ioctl_completion->overlapped.hEvent != nullptr)
+            CloseHandle(async_ioctl_completion->overlapped.hEvent);
+
+        ebpf_free(async_ioctl_completion);
+    }
+}
+
+OVERLAPPED*
+get_async_ioctl_operation_overlapped(_In_ const async_ioctl_completion_t* async_ioctl_completion)
+{
+    return const_cast<OVERLAPPED*>(&async_ioctl_completion->overlapped);
+}
+
+ebpf_result_t
+get_async_ioctl_result(_In_ const async_ioctl_completion_t* ioctl_completion)
+{
+    DWORD dummy;
+    if (!GetOverlappedResult(
+            reinterpret_cast<HANDLE>(get_device_handle()),
+            get_async_ioctl_operation_overlapped(ioctl_completion),
+            &dummy,
+            FALSE))
+        return win32_error_code_to_ebpf_result(GetLastError());
+    return EBPF_SUCCESS;
+}
+
+ebpf_result_t
+initialize_async_ioctl_operation(
+    _Inout_opt_ void* callback_context,
+    _In_ const async_ioctl_completion_callback_t callback,
+    _Outptr_ async_ioctl_completion_t** async_ioctl_completion)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    *async_ioctl_completion = NULL;
+
+    async_ioctl_completion_context_t* local_async_ioctl_completion =
+        (async_ioctl_completion_context_t*)ebpf_allocate(sizeof(async_ioctl_completion_context_t));
+    if (local_async_ioctl_completion == nullptr) {
+        result = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+
+    local_async_ioctl_completion->callback_context = callback_context;
+    local_async_ioctl_completion->callback = callback;
+
+    // Create Event object for OVERLAPPED struct.
+    local_async_ioctl_completion->overlapped.hEvent = CreateEvent(nullptr, false, false, nullptr);
+    if (local_async_ioctl_completion->overlapped.hEvent == nullptr) {
+        result = win32_error_code_to_ebpf_result(GetLastError());
+        _Analysis_assume_(result != EBPF_SUCCESS);
+        EBPF_LOG_WIN32_API_FAILURE(EBPF_TRACELOG_KEYWORD_API, CreateEvent);
+        goto Exit;
+    }
+
+    // Setup threadpool wait for the overlapped hEvent and pass the async completion context as wait callback context.
+    local_async_ioctl_completion->wait = CreateThreadpoolWait(
+        [](_In_ const PTP_CALLBACK_INSTANCE instance,
+           _Inout_ void* context,
+           _Inout_ PTP_WAIT wait,
+           TP_WAIT_RESULT wait_result) {
+            UNREFERENCED_PARAMETER(instance);
+            UNREFERENCED_PARAMETER(wait);
+            UNREFERENCED_PARAMETER(wait_result);
+
+            async_ioctl_completion_context_t* local_async_ioctl_completion = (async_ioctl_completion_context_t*)context;
+            local_async_ioctl_completion->callback(local_async_ioctl_completion->callback_context);
+        },
+        local_async_ioctl_completion,
+        NULL);
```

Comment:
```suggestion
        nullptr);
```
in cpp files, use nullptr instead of NULL except where the compiler would complain about nullptr